### PR TITLE
XOR is not a hash function

### PR DESCRIPTION
**Pull request by SchumBlubBlub**
piccard at tanagra, weeping.

### CODE REVIEW

**Comment by condret** [comment]
thanks for your pr, but it was intended that this code uses sdb anyway. If you want you could do that

File: libr/flag/flag.c [hunk 1]
```diff
@@ -389,7 +409,7 @@ R_API char *r_flag_get_liststr(RFlag *f, ut64 off) {
 	char *p = NULL;
 	r_list_foreach (list, iter, fi) {
 		p = r_str_concatf (p, "%s%s",
-			fi->realname, iter->n ? "," : ":");
+				fi->realname, iter->n? ",": ":");
```
Comment (radare) [inline libr/flag/flag.c]: autoindentation fail

File: libr/flag/flag.c [hunk 2]
```diff
@@ -624,11 +644,13 @@ int main () {
 	r_flag_set (f, "rip", 4, 4);
 
 	i = r_flag_get (f, "rip");
-	if (i) printf ("nRIP: %p %llx\n", i, i->offset);
+	if (i)
+		printf ("nRIP: %p %llx\n", i, i->offset);
```
Comment (radare) [inline libr/flag/flag.c]: use PFMT64x not llx

File: libr/flag/flag.c [hunk 3]
```diff
@@ -624,11 +644,13 @@ int main () {
...
 	else printf ("nRIP: null\n");
 
 	i = r_flag_get_i (f, 0xfff333999000LL);
-	if (i) printf ("iRIP: %p %llx\n", i, i->offset);
+	if (i)
+		printf ("iRIP: %p %llx\n", i, i->offset);
```
Comment (radare) [inline libr/flag/flag.c]: use pfmt64x and braces

File: libr/flag/flag.c [hunk 4]
```diff
@@ -1,29 +1,50 @@
...
-#define XOROFF(x) (x^XORKEY)
+R_LIB_VERSION (r_flag);
 
 #define ISNULLSTR(x) (!(x) || !*(x))
 #define IS_IN_SPACE(f, i) ((f)->space_idx != -1 && (i)->space != (f)->space_idx)
 
+// Sources: http://burtleburtle.net/bob/hash/integer.html
+//          http://www.cris.com/~Ttwang/tech/inthash.htm
+// Described as public domain
+static ut32 hash_func_32(ut32 a) {
```
Comment (radare) [inline libr/flag/flag.c]: there's no need to hash numbers anymore, we have collision free hashtables and sdb, which was not the case at the time this file was written.

Flags should be changed to use Sdb or SdbHt.

**Comment by Maijin** [comment]
See https://github.com/radare/radare2/issues/2761#issuecomment-278802440

File: libr/flag/flag.c (hunk 4 above)
Comment (SchumBlubBlub) [inline libr/flag/flag.c]: I don't know all the detail but using XOR as hash func might be *why* you had lots of collisions.

**Comment by radare** [comment]
nope, the issue happened when not using any hash, it is a hack, it works but its a hack and can be completelly ignored if using SdbHt or Sdb to store all the flags instead of the old ht.c this was part of our refactoring roadmap, as well as reviewing the API of r_flags and try to determine if there are functions that can be removed or renamed to clarify its use.

additionally in r2r we have a wip unit test that should have tests for r_flags too using the API instead of r2 commands.

File: libr/flag/flag.c (hunk 2 above)
Comment (SchumBlubBlub) [inline libr/flag/flag.c]: I ran sys/indent and it changed the whole file. I'm nowhere near this function, I can't fix everything because of that.

**Comment by SchumBlubBlub** [comment]
Ok, where can I see the roadmap?

**Pull request closed without merging**